### DISCUSSION TITLE
Running grid_stat on multiple models

### DISCUSSION THREAD

**Discussion by lesnow93** [dtcenter/METplus#1838] [Configuration] [answered]
Hello METplus forum!

I'm looking to run METplus with pcp_combine+grid_stat on multiple individual model runs in order to replicate previous grid_stat results using MET and my own wrapper scripts outside of METplus.  After that, I plan to process the multi-model system through ensemble_stat to verify probabilities.  So I'd like to know if a METplus conf file template can be created to handle the processing of multiple model runs into one job that runs grid_stat (similar to using a wildcard in the EnsembleStat.conf use case), or do I have to re-run METplus for pcp_combine+grid_stat for each individual model run?  I haven't yet come across such an example or use-case.

Specifically, can the following be made into lists?
MODEL =MODEL1, MODEL2, MODEL3, etc.
FCST_GRID_STAT_INPUT_TEMPLATE= MODEL1_{init?fmt=%y%m%d%H}00.grb2f{lead?fmt=%HH}0000,
  MODEL2_{init?fmt=%y%m%d%H}00.grb2f{lead?fmt=%HH}0000, 
  MODEL3_{init?fmt=%y%m%d%H}00.grb2f{lead?fmt=%HH}0000, 
  etc.

Thanks!
-Jonathan

**Comment by georgemccabe** [comment, marked as answer]
Hi Jonathan,

Yes, there are a few ways to configure METplus to handle what you need. Custom looping should accomplish what you are describing in your example. Wrapper instances can be used for more complex examples. Both are described below. Let me know if you have any other questions.

## Custom Loop List

If the only difference between each of your runs is a single string, then you can use the custom loop list logic to define a list of strings that can be referenced and substituted for multiple runs of a wrapper. It is described in the METplus User's Guide in the Configuration chapter in the [Custom Looping](https://metplus.readthedocs.io/en/latest/Users_Guide/systemconfiguration.html#custom-looping) section.

Here is a PCPCombine use case that utilizes custom loops:

https://metplus.readthedocs.io/en/latest/generated/met_tool_wrapper/PCPCombine/PCPCombine_loop_custom.html#metplus-configuration

In your case, you could set:

`GRID_STAT_CUSTOM_LOOP_LIST = MODEL1, MODEL2, MODEL3`

and reference each value using {custom?fmt=%s}:

`FCST_GRID_STAT_INPUT_TEMPLATE = {custom?fmt=%s}_{init?fmt=%y%m%d%H}00.grb2f{lead?fmt=%HH}0000`

I believe you can simply use `{custom}` instead of provided the format info.

Each runtime will call GridStat 3 times, once for each item in the custom loop list.

## Wrapper Instances

If you need to run a wrapper multiple times with different settings for each, then you can define an instance identifier string. This is described in the METplus User's Guide in the Configuration Chapter in the [Process List](https://metplus.readthedocs.io/en/latest/Users_Guide/systemconfiguration.html#process-list) (see Example 3).

```
[config]

PROCESS_LIST = GridStat(MODEL1), GridStat(MODEL2), GridStat(MODEL3)

[MODEL1]

FCST_GRID_STAT_INPUT_TEMPLATE= MODEL1_{init?fmt=%y%m%d%H}00.grb2f{lead?fmt=%HH}0000

[MODEL2]

FCST_GRID_STAT_INPUT_TEMPLATE= MODEL2_{init?fmt=%y%m%d%H}00.grb2f{lead?fmt=%HH}0000

[MODEL3]

FCST_GRID_STAT_INPUT_TEMPLATE= MODEL3_{init?fmt=%y%m%d%H}00.grb2f{lead?fmt=%HH}0000
```

In this example, there are 3 instances of GridStat defined in the PROCESS_LIST and there are config section headings that match the instance identifier string. Settings defined in these sections will override any settings found in the [config] section for that instance only.

In this example, the only difference between the 3 instances is the instance string, so you could use the `{instance}` tag to substitute the instance string instead of defining variables in each section. The following will accomplish the same behavior:

```
[config]

PROCESS_LIST = GridStat(MODEL1), GridStat(MODEL2), GridStat(MODEL3)

FCST_GRID_STAT_INPUT_TEMPLATE= {instance}_{init?fmt=%y%m%d%H}00.grb2f{lead?fmt=%HH}0000
```

**Reply by lesnow93** [reply]
Thanks for this detailed reply, George!  I believe either solution will work for what I need to accomplish, but the former solution is probably the cleaner, more concise pathway.  I will need to run both pcp_combine and grid_stat in my wrapper config, so will I need to set up custom looping for both of these processes using the {custom?fmt=%s}?  

i.e., 
PCP_COMBINE_CUSTOM_LOOP_LIST = MODEL1, MODEL2, MODEL3, etc.
and
GRID_STAT_CUSTOM_LOOP_LIST = MODEL1, MODEL2, MODEL3, etc.

Thanks much!
-Jonathan

**Reply by georgemccabe** [reply]
Hi Jonathan,

You can define a custom loop list that applies to all wrappers using [CUSTOM_LOOP_LIST](https://metplus.readthedocs.io/en/latest/Users_Guide/glossary.html#term-CUSTOM_LOOP_LIST).

**Reply by lesnow93** [reply]
Thanks again.  That's even simpler and better!
-Jonathan

**Reply by georgemccabe** [reply]
Great! I'm going to mark this discussion as answered and lock it to prevent future posts. Please create a new discussion if you have any other questions. Thanks!
